proposals/0030-core-image-story.md

```diff
@@ -0,0 +1,430 @@
+<Image src={myImage} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/my_image.hash.webp"
+  width="1600"
+  height="900"
+  decoding="async"
+  loading="lazy"
+  alt="..."
+/>
+```
+
+### Remote images
+
+```astro
+---
+import { Image } from "astro:assets";
+// Image is 1920x1080
+---
+
+<Image src="https://example.com/image.png" alt="..." />
+<!-- ERROR! `width` and `height` are required -->
+
+<Image src="https://example.com/image.png" width={1280} alt="..." />
+<!-- ERROR! `height` is required -->
+
+<Image src="https://example.com/image.png" width={1280} height={720} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Remote images are not optimized or resized. Nonetheless, you still benefit from the enforced attributes and the guarantee of no CLS, albeit manually. -->
+<img
+  src="https://example.com/image.png"
+  decoding="async"
+  loading="lazy"
+  width="1280"
+  height="720"
+  alt="..."
+/>
+```
+
+## Image in Markdown
+
+```markdown
+![My article cover](./cover.png)
+^ Image is 1280x720
+```
+
+### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/cover.hash.webp"
+  width="1280"
+  height="720"
+  decoding="async"
+  loading="lazy"
+  alt="My article cover"
+/>
+```
+
+Remote images are handled as in vanilla Markdown and have no special behavior.
+
+## JavaScript API for using optimized images
+
+```astro
+---
+import { getImage } from "astro:assets";
+import image from "../assets/image.png"; // Image is 1920x1080
+
+const myImage = await getImage({ src: image, width: 1280, height: 720 });
+---
+
+<img src={myImage.src} {...myImage.attributes} />
+```
+
+### Result
+
+```html
+<img
+  src="/_astro/image.hash.webp"
+  width="1280"
+  height="720"
+  loading="lazy"
+  decoding="async"
+/>
+```
+
+## Defining a new image service
+
+### Astro Config
+
+```ts
+import { defineConfig } from "astro/config";
+
+// https://astro.build/config
+export default defineConfig({
+  image: {
+    service: "your-entrypoint", // 'astro/image/services/squoosh' | 'astro/image/services/sharp' | string
+  },
+});
+```
+
+### Local services API
+
+```ts
+import type { LocalImageService } from "astro";
+
+const service: LocalImageService = {
+  getURL(options: ImageTransform) {
+    return `/my_super_endpoint_that_transforms_images?w=${options.width}`;
+  },
+  parseURL(url: URL) {
+    return {
+      width: url.searchParams.get("w"),
+    };
+  },
+  transform(options: ImageTransform) {
+    const { buffer } = mySuperLibraryThatEncodesImages(options);
+
+    return {
+      data: buffer,
+      format: options.format,
+    };
+  },
+};
+
+export default service;
+```
+
+### External services API
+
+```ts
+import type { ExternalImageService } from "astro";
+
+const service: ExternalImageService = {
+  getURL(options: ImageTransform) {
+    return `https://mysupercdn.com/${options.src}?q=${options.quality}`;
+  },
```

Comment:
This doesn't seem to include the path to the image. Is the (fingerprinted) path part of `options`?

Reply:
Sorry, I'm not sure I get what you mean. This is a remote service, there's no hashed files here

Reply:
> there's no hashed files here

I think that's a failure. Specifically, I envision the following course of events:

1. Developer configures an `ExternalImageService` per your example
2. Developer adds a new `src/images/foo.png` and builds site. Published site refers to `<img src="https://mysupercdn.com/_astro/images/foo.png?q=80">`
3. User loads site. Browser makes request to `https://mysupercdn.com/_astro/images/foo.png?q=80` to fetch the optimized version.
4. MySuperCDN makes request to `https://myorigin.com/_astro/images/foo.png` to fetch the original version. Origin responds. MySuperCDN caches the original, performs transforms, caches the transforms
5. Developer updates `src/images/foo.png` and builds site. Published site still refers to `<img src="https://mysupercdn.com/_astro/images/foo.png?q=80">`

There are three possibilities, depending on configuration

1. If the origin uses a long cache time for `foo.png` (e.g. `cache-control: public, max-age=31536000`), then this will result in MySuperCDN having outdated assets. Performance will be good, but the asset will never get cache-busted.
2. If the origin uses a short cache time for `foo.png` (e.g. `cache-control: public, max-age=60`), then this will result in MySuperCDN having to refetch `foo.png` and re-perform transforms frequently, hurting site performance. It also means the asset will be outdated for 60 seconds after a deploy.
3. Some image-transforms CDNs allow content uploads. In this case, the site would have an Astro integration that pushes images to the image service. The service would know that the image has changed and would expire its caches. This won't work with services like Imgix, Fastly, or Cloudflare because their image optimizers all work as caching reverse proxies. It would work with Cloudinary, which has a Digital Asset Management product attached to its image optimizer. But in the Cloudinary case, it would introduce a new failure mode: the image asset would be updated before the deploy finishes. If the deploy fails, you now have old HTML with the new image.

For any of these cases, the solution is simple: pass the fingerprinted path to the image service.

Reply:
Hm, I see what you mean. Do you know how other tools help with this? None of the other tools with this kind of API seems to offer a fingerprint, as far as I could tell. 

I'm not sure either how we would offer this, as kinda no fingerprint exists for the original version either.

Reply:
Ah, I had misunderstood what you meant initially. If you pass an ESM imported image, you'll get a fingerprinted path, yes.